### PR TITLE
Type Lowering and Address Types

### PR DESCRIPTION
**Pull request by CodaFi**
### What's in this pull request?

Implement a centralized type lowering facility and make it distinguish objects and addresses.

<strike>WIP because there's no IRGen and because the lowerings should become the authority on how to operate on their values.</strike>

### CODE REVIEW

File: Tests/Mesosphere/dispatch-nested.silt [hunk 1]
```diff
@@ -63,6 +66,7 @@ doThing (ascii b) = b
 -- CHECK-GIR:   %15 = data_init dispatch-nested.Bool ; dispatch-nested.Bool.false
 -- CHECK-GIR:   %16 = data_init dispatch-nested.Bool ; dispatch-nested.Bool.false
 -- CHECK-GIR:   %17 = data_init dispatch-nested.Byte ; dispatch-nested.Byte.byte ; %9 ; %10 ; %11 ; %12 ; %13 ; %14 ; %15 ; %16
+-- CHECK-GIR:   destroy_value %0
```
Comment (CodaFi) [inline Tests/Mesosphere/dispatch-nested.silt]: Whoopsie

**Comment by harlanhaskins** [comment]
I can handle the IRGen phase.

**Review by harlanhaskins — APPROVED** [review]
This is really great!

Can you please add doc comments above each op detailing their intended semantics? That would really help with IRGen.

**Comment by CodaFi** [comment]
⛵️

**Pull request merged**